tlsn/tlsn-core/src/proof.rs

```diff
@@ -0,0 +1,251 @@
+//! Types for substring proofs.
+
+use std::collections::HashMap;
+
+use mpz_circuits::types::ValueType;
+use serde::{Deserialize, Serialize};
+use utils::range::{RangeDisjoint, RangeSet, RangeUnion};
+
+use crate::{
+    commitment::{Commitment, CommitmentId, CommitmentInfo, CommitmentOpening},
+    merkle::MerkleProof,
+    transcript::get_encoding_ids,
+    Direction, EncodingId, SessionData, SessionHeader, TranscriptSlice, MAX_TOTAL_COMMITTED_DATA,
+};
+
+use mpz_garble_core::Encoder;
+
+/// An error for [`SubstringsProofBuilder`]
+#[derive(Debug, thiserror::Error)]
+#[non_exhaustive]
+pub enum SubstringsProofBuilderError {
+    /// Invalid commitment id.
+    #[error("invalid commitment id: {0:?}")]
+    InvalidCommitmentId(CommitmentId),
+    /// Invalid commitment type.
+    #[error("commitment {0:?} is not a substrings commitment")]
+    InvalidCommitmentType(CommitmentId),
+    /// Attempted to add a commitment with a duplicate id.
+    #[error("commitment with id {0:?} already exists")]
+    DuplicateCommitmentId(CommitmentId),
+}
+
+/// A builder for [`SubstringsProof`]
+pub struct SubstringsProofBuilder<'a> {
+    data: &'a SessionData,
+    openings: HashMap<CommitmentId, (CommitmentInfo, CommitmentOpening)>,
+}
+
+opaque_debug::implement!(SubstringsProofBuilder<'_>);
+
+impl<'a> SubstringsProofBuilder<'a> {
+    /// Creates a new builder.
+    pub(crate) fn new(data: &'a SessionData) -> Self {
+        Self {
+            data,
+            openings: HashMap::default(),
+        }
+    }
+
+    /// Reveals data corresponding to the provided commitment id
+    pub fn reveal(&mut self, id: CommitmentId) -> Result<&mut Self, SubstringsProofBuilderError> {
+        let commitment = self
+            .data
+            .commitments()
+            .get(&id)
+            .ok_or(SubstringsProofBuilderError::InvalidCommitmentId(id))?;
+
+        let info = self
+            .data
+            .commitments()
+            .get_info(&id)
+            .expect("info exists if commitment exists");
+
+        #[allow(irrefutable_let_patterns)]
+        let Commitment::Blake3(commitment) = commitment
+        else {
+            return Err(SubstringsProofBuilderError::InvalidCommitmentType(id));
+        };
+
+        let transcript = match info.direction() {
+            Direction::Sent => self.data.sent_transcript(),
+            Direction::Received => self.data.recv_transcript(),
+        };
+
+        let data = transcript.get_bytes_in_ranges(info.ranges());
+
+        // check that the commitment is not already revealed
```

Comment:
Comment is a little bit mislading, better:
```suggestion
        // add commitment to openings and return an error if it is already present
```